File: internal/harness/copilot.go
package harness

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/jrunic/koine/internal/cache"
	"github.com/jrunic/koine/internal/render"
)

// Copilot é o adapter do harness para o GitHub Copilot CLI.
//
// Materializa em ~/.cache/koine/copilot-bundles/<slot>/:
//   - AGENTS.md — USUARIO + Hermes mesclados (frontmatter removido, H1→H2)
//   - .github/instructions/escopo.instructions.md
//   - .github/instructions/kn-indice-<dom>.instructions.md (por entrada em IndicePaths)
//
// Cria symlink <pasta>/.github/copilot-instructions.md → <pasta>/CONTEXTO.md.
// Seta COPILOT_CUSTOM_INSTRUCTIONS_DIRS apontando para o bundle.
// Modo bootstrap: apenas AGENTS.md + env var; sem symlink nem instructions.
type Copilot struct {
	// Agente é o nome da persona canônica (ex: "hermes").
	Agente string
	// PastaAbs é o diretório de trabalho absoluto — determina SlotID e alvo do symlink.
	PastaAbs string
}

func (c *Copilot) Nome() string {
	return "copilot"
}

func (c *Copilot) CaminhoArquivoContexto(cwd string) string {
	return filepath.Join(cwd, ".github", "copilot-instructions.md")
}

func (c *Copilot) Renderizar(dados ContextoMontado) (Lancamento, error) {
	slotID := cache.SlotID(c.PastaAbs)
	bundleDir := cache.CaminhoBundle("copilot-bundles", slotID)

	lancamento := Lancamento{
		ArquivosExternos: make(map[string][]byte),
		EnvVars: map[string]string{
			"COPILOT_CUSTOM_INSTRUCTIONS_DIRS": bundleDir,
		},
	}

	agentsMD, err := c.montarAgentsMD(dados)
	if err != nil {
		return Lancamento{}, err
	}
	lancamento.ArquivosExternos[filepath.Join(bundleDir, "AGENTS.md")] = agentsMD

	if dados.Bootstrap {
		if dados.ContextoPath != "" {
			conteudo, err := os.ReadFile(dados.ContextoPath)
			if err != nil {
				return Lancamento{}, fmt.Errorf("lendo CONTEXTO.md em bootstrap: %w", err)
			}
			instrPath := filepath.Join(bundleDir, ".github", "instructions", "bootstrap.instructions.md")
			lancamento.ArquivosExternos[instrPath] = render.WraparInstructions(conteudo)
		}
		return lancamento, nil
	}

	if dados.EscopoPath != "" {
		conteudo, err := os.ReadFile(dados.EscopoPath)
		if err != nil {
			return Lancamento{}, fmt.Errorf("lendo escopo: %w", err)
		}
		instrPath := filepath.Join(bundleDir, ".github", "instructions", "escopo.instructions.md")
		lancamento.ArquivosExternos[instrPath] = render.WraparInstructions(conteudo)
	}

	for _, indicePath := range dados.IndicePaths {
		conteudo, err := os.ReadFile(indicePath)
		if err != nil {
			return Lancamento{}, fmt.Errorf("lendo índice %s: %w", indicePath, err)
		}
		dom := dominioDe(indicePath)
		instrPath := filepath.Join(bundleDir, ".github", "instructions", "kn-indice-"+dom+".instructions.md")
		lancamento.ArquivosExternos[instrPath] = render.WraparInstructions(conteudo)
	}

	linkPath := filepath.Join(c.PastaAbs, ".github", "copilot-instructions.md")
	lancamento.Symlinks = map[string]string{
		linkPath: dados.ContextoPath,
	}

	return lancamento, nil
}

func (c *Copilot) montarAgentsMD(dados ContextoMontado) ([]byte, error) {
	var partes []render.Parte

	if dados.UsuarioPath != "" {
		conteudo, err := os.ReadFile(dados.UsuarioPath)
		if err != nil {
			return nil, fmt.Errorf("lendo usuario: %w", err)
		}
		partes = append(partes, render.Parte{Secao: "Usuário", Conteudo: conteudo})
	}

	conteudoAgente, err := os.ReadFile(dados.AgentePath)
	if err != nil {
		return nil, fmt.Errorf("lendo agente %s: %w", dados.AgentePath, err)
	}
	partes = append(partes, render.Parte{Secao: "Agente", Conteudo: conteudoAgente})

	return render.MescarDocumentos("Sessão Koine — Copilot", partes), nil
}

// dominioDe extrai o nome de domínio de um caminho de índice.
// Ex: /foo/kn-indice-universal.md → "universal"
func dominioDe(indicePath string) string {
	base := strings.TrimSuffix(filepath.Base(indicePath), ".md")
	return strings.TrimPrefix(base, "kn-indice-")
}

// Verifica em compile-time que Copilot satisfaz Harness.
var _ Harness = (*Copilot)(nil)

File: src/koine/adapters/copilot.py
import os

from koine import cache, render, shell
from koine import paseo as _paseo
from koine.contexto import ContextoMontado
from koine.lancamento import Lancamento

# Arquivo do Copilot na pasta do usuário — hoje só no `gerar` (modo skills). No
# launch a pasta não recebe nada: o bundle entrega tudo.
ARQUIVO = os.path.join(".github", "copilot-instructions.md")
MARCADOR = "<!-- gerado por kn-agente -->"

# Medido em 28/08/2026: usa PowerShell e SO PowerShell no Windows, e nao expoe
# chave para trocar. Onde a politica nega o powershell.exe, nao tem shell.
ACEITA_SHELL = (shell.PWSH, shell.POWERSHELL)

# Rota pelo Paseo, medida em 29/08/2026: mesma forma do claude — o Paseo o
# invoca com `--acp`, do cwd do workspace.
PASEO = _paseo.Rota(extends="copilot")


def renderizar(cm: ContextoMontado) -> Lancamento:
    """Porta de harness.Copilot.Renderizar (copilot.go).

    Materializa o bundle em ~/.cache/koine/copilot-bundles/<slot>/, com TODAS
    as camadas em .github/instructions/*.instructions.md. Cria symlink
    <pasta>/.github/copilot-instructions.md → CONTEXTO.md e seta
    COPILOT_CUSTOM_INSTRUCTIONS_DIRS. Bootstrap: usuário + agente + env
    (+ bootstrap.instructions.md se houver CONTEXTO.md); sem symlink.

    O KOINE.md entra aqui desde a v0.7.1: o bundle nunca o levou (herança do
    copilot.go), e sem ele o agente sabia quem é o usuário e em que pasta está,
    mas não sabia o que é o Koine — em 3682 bytes, ~3% do input da sessão.

    O bundle NÃO tem AGENTS.md: medido em 27/08 com discriminante na mesma
    execução, o canal COPILOT_CUSTOM_INSTRUCTIONS_DIRS entrega os
    `*.instructions.md` e ignora o AGENTS.md. Usuário e agente moravam nele —
    isto é, não chegavam à sessão. Arquivo que o cliente não lê só engana quem
    for depurar."""
    bundle = cache.caminho_bundle("copilot-bundles", cache.slot_id(cm.pasta_abs))
    instr = os.path.join(bundle, ".github", "instructions")
    lanc = Lancamento(env_vars={"COPILOT_CUSTOM_INSTRUCTIONS_DIRS": bundle})

    if cm.usuario_path:
        lanc.arquivos_externos[os.path.join(instr, "usuario.instructions.md")] = \
            render.wrapar_instructions(_ler(cm.usuario_path))
    if cm.koine_path:
        lanc.arquivos_externos[os.path.join(instr, "koine.instructions.md")] = \
            render.wrapar_instructions(_ler(cm.koine_path))
    lanc.arquivos_externos[os.path.join(instr, "agente.instructions.md")] = \
        render.wrapar_instructions(_ler(cm.agente_path))

    if cm.instrucao_path:
        lanc.arquivos_externos[os.path.join(instr, "instrucao.instructions.md")] = \
            render.wrapar_instructions(_ler(cm.instrucao_path))

    # O CONTEXTO.md vai por CONTEÚDO, no bundle. Antes chegava por symlink na
    # pasta do usuário — e symlink na pasta é justamente o que sai de cena.
    if cm.contexto_path:
        lanc.arquivos_externos[os.path.join(instr, "contexto.instructions.md")] = \
            render.wrapar_instructions(_ler(cm.contexto_path))
    if cm.bootstrap:
        return lanc

    if cm.escopo_path:
        lanc.arquivos_externos[os.path.join(instr, "escopo.instructions.md")] = \
            render.wrapar_instructions(_ler(cm.escopo_path))
    for ip in cm.indice_paths:
        dom = render.dominio_de(ip)
        lanc.arquivos_externos[os.path.join(instr, f"kn-indice-{dom}.instructions.md")] = \
            render.wrapar_instructions(_ler(ip))

    return lanc


def renderizar_para_pasta(cm: ContextoMontado) -> tuple[str, str]:
    """Materialização a pedido (`koine gerar`, modo skills).

    Aqui o conteúdo vai INLINE: sem wrapper não há
    COPILOT_CUSTOM_INSTRUCTIONS_DIRS para apontar o bundle, e o
    `.github/copilot-instructions.md` é a única via que sobra.
    """
    return ARQUIVO, MARCADOR + "\n" + render.documento_inline(
        "Sessão Koine — Copilot", cm) + "\n"


def _ler(path: str) -> str:
    with open(path, encoding="utf-8") as f:
        return f.read()
